File: mobile/lib/core/models/scoring.dart
import 'form_template.dart';

/// Calcul note -> pourcentage -> mention, réimplémentation Dart de
/// `shared/src/scoring.ts` à partir du même modèle [ConversionTable].
///
/// Différence assumée avec la version TypeScript (pensée pour une
/// section entièrement remplie, côté backend) : ici, les critères non
/// encore notés comptent pour 0 (au lieu d'être exclus du calcul), afin
/// d'afficher un score cumulé qui progresse en temps réel pendant la
/// saisie plutôt que de rester artificiellement élevé tant que la section
/// n'est pas terminée. Les deux implémentations convergent vers le même
/// résultat une fois tous les critères renseignés.
class ConversionTableError extends StateError {
  ConversionTableError(super.message);
}

class ConversionResult {
  const ConversionResult({required this.scoreOn4, required this.percentage, required this.mention});

  final int scoreOn4;
  final double percentage;
  final String mention;
}

class SectionScoreResult {
  const SectionScoreResult({
    required this.totalScore,
    required this.maxScore,
    required this.percentage,
    required this.mention,
    required this.scoreOn4,
    required this.answeredCount,
    required this.criteriaCount,
  });

  final int totalScore;
  final int maxScore;
  final double percentage;
  final String mention;
  final int scoreOn4;
  final int answeredCount;
  final int criteriaCount;

  bool get isComplete => answeredCount >= criteriaCount;
}

/// Convertit une note brute en pourcentage + mention, via le "Tableau de
/// conversion" du formulaire.
///
/// [criteriaCount] est requis en mode `lookupByCriteriaCount` : le nombre
/// de critères concernés (section), ou le nombre de sections pour la
/// synthèse finale.
ConversionResult convertRawScore({
  required ConversionTable table,
  required int rawScore,
  required int maxScore,
  int? criteriaCount,
}) {
  final percentage = maxScore == 0 ? 0.0 : ((rawScore / maxScore) * 10000).roundToDouble() / 100;

  if (table.mode == ConversionTableMode.percentageOnly) {
    for (final band in table.bands) {
      if (percentage >= band.minPercentage && percentage <= band.maxPercentage) {
        return ConversionResult(scoreOn4: band.scoreOn4, percentage: percentage, mention: band.mention);
      }
    }
    throw ConversionTableError('Aucune bande de conversion ne couvre $percentage%.');
  }

  final rows = table.rows;
  if (criteriaCount == null || rows == null) {
    throw ArgumentError('criteriaCount et conversionTable.rows sont requis en mode lookup_by_criteria_count.');
  }
  ConversionTableRow? row;
  for (final candidate in rows) {
    if (candidate.criteriaCount == criteriaCount) {
      row = candidate;
      break;
    }
  }
  if (row == null) {
    throw ConversionTableError('Aucune ligne du tableau de conversion pour $criteriaCount critères.');
  }
  for (var i = 0; i < row.ranges.length && i < table.bands.length; i++) {
    if (row.ranges[i].contains(rawScore)) {
      final band = table.bands[i];
      return ConversionResult(scoreOn4: band.scoreOn4, percentage: percentage, mention: band.mention);
    }
  }
  throw ConversionTableError('Aucune plage ne couvre la note $rawScore pour $criteriaCount critères.');
}

/// Calcule le score d'une section à partir des notes déjà saisies
/// (`scores`, indexé par identifiant de critère — les critères absents de
/// la map comptent pour 0).
SectionScoreResult computeSectionScore({
  required FormSectionTemplate section,
  required Map<String, int> scores,
  required ConversionTable conversionTable,
}) {
  var totalScore = 0;
  var maxScore = 0;
  var answeredCount = 0;

  for (final criterion in section.criteria) {
    final score = scores[criterion.id];
    if (score != null) {
      answeredCount++;
      totalScore += score;
    }
    maxScore += criterion.maxScore;
  }

  final result = convertRawScore(
    table: conversionTable,
    rawScore: totalScore,
    maxScore: maxScore,
    criteriaCount: section.criteria.length,
  );

  return SectionScoreResult(
    totalScore: totalScore,
    maxScore: maxScore,
    percentage: result.percentage,
    mention: result.mention,
    scoreOn4: result.scoreOn4,
    answeredCount: answeredCount,
    criteriaCount: section.criteria.length,
  );
}

/// Calcule l'évaluation synthétique finale à partir des scores de chaque
/// section (`sectionScores`, indexé par identifiant de section), en
/// reconvertissant la somme des notes-sur-4 via le même tableau de
/// conversion — voir `SynthesisTemplate.conversionCriteriaCount`.
ConversionResult computeSynthesisScore({
  required FormTemplate template,
  required Map<String, SectionScoreResult> sectionScores,
}) {
  var totalScoreOn4 = 0;
  final rowCount = template.synthesis.rows.length;

  for (final row in template.synthesis.rows) {
    totalScoreOn4 += sectionScores[row.sectionId]?.scoreOn4 ?? 0;
  }

  final maxScore = rowCount * 4;
  final criteriaCount = template.synthesis.conversionCriteriaCount ?? rowCount;

  return convertRawScore(
    table: template.conversionTable,
    rawScore: totalScoreOn4,
    maxScore: maxScore,
    criteriaCount: criteriaCount,
  );
}

File: shared/src/scoring.ts
import { CriterionResponse } from './types/form-submission';
import {
  ConversionTable,
  FormSectionTemplate,
} from './types/form-template';

/**
 * Logique de calcul partagée entre le backend, le web et (via le même
 * algorithme réimplémenté en Dart) le mobile : conversion des notes brutes
 * d'une section en pourcentage puis en mention, à partir du "Tableau de
 * conversion" unique défini au niveau du `FormTemplate`.
 */

export class UnknownCriterionError extends Error {
  constructor(criterionId: string, sectionId: string) {
    super(`Critère inconnu "${criterionId}" pour la section "${sectionId}".`);
    this.name = 'UnknownCriterionError';
  }
}

export class NoMatchingConversionRuleError extends Error {
  constructor(criteriaCount: number, rawScore: number) {
    super(
      `Aucune règle du tableau de conversion ne couvre la note ${rawScore} pour ${criteriaCount} critères. ` +
        'Vérifiez que conversionTable.rows contient une ligne pour ce nombre de critères et que ses plages couvrent 0-(4×N) sans trou.',
    );
    this.name = 'NoMatchingConversionRuleError';
  }
}

export interface ConversionResult {
  scoreOn4: 4 | 3 | 2 | 1 | 0;
  percentage: number;
  mention: string;
}

/**
 * Convertit une note brute en pourcentage + mention, via le "Tableau de
 * conversion" du formulaire.
 *
 * @param criteriaCount Nombre de critères concernés (N) — la longueur de
 *   `FormSectionTemplate.criteria`, ou `SynthesisTemplate.conversionCriteriaCount`
 *   pour la synthèse finale. Ignoré en mode `percentage_only`.
 */
export function convertRawScore(
  table: ConversionTable,
  rawScore: number,
  maxScore: number,
  criteriaCount?: number,
): ConversionResult {
  const percentage = maxScore === 0 ? 0 : Math.round((rawScore / maxScore) * 10000) / 100;

  if (table.mode === 'percentage_only') {
    const band = table.bands.find(
      (b) => percentage >= b.minPercentage && percentage <= b.maxPercentage,
    );
    if (!band) {
      throw new NoMatchingConversionRuleError(criteriaCount ?? 0, rawScore);
    }
    return { scoreOn4: band.scoreOn4, percentage, mention: band.mention };
  }

  if (criteriaCount === undefined) {
    throw new Error('criteriaCount est requis en mode lookup_by_criteria_count.');
  }
  const row = table.rows?.find((r) => r.criteriaCount === criteriaCount);
  if (!row) {
    throw new NoMatchingConversionRuleError(criteriaCount, rawScore);
  }
  const bandIndex = row.ranges.findIndex((r) => rawScore >= r.min && rawScore <= r.max);
  if (bandIndex === -1) {
    throw new NoMatchingConversionRuleError(criteriaCount, rawScore);
  }
  const band = table.bands[bandIndex];
  return { scoreOn4: band.scoreOn4, percentage, mention: band.mention };
}

export interface SectionScoreResult {
  totalScore: number;
  maxScore: number;
  percentage: number;
  mention: string;
  /** Note de la section convertie sur 4, réutilisable pour l'évaluation synthétique finale. */
  scoreOn4?: 4 | 3 | 2 | 1 | 0;
}

/**
 * Calcule le score total, le pourcentage et la mention d'une section à
 * partir des réponses aux critères et du tableau de conversion du template.
 */
export function computeSectionScore(
  section: FormSectionTemplate,
  responses: CriterionResponse[],
  conversionTable: ConversionTable,
): SectionScoreResult {
  const criteriaById = new Map(section.criteria.map((c) => [c.id, c]));

  let totalScore = 0;
  let maxScore = 0;

  for (const response of responses) {
    const criterion = criteriaById.get(response.criterionId);
    if (!criterion) {
      throw new UnknownCriterionError(response.criterionId, section.id);
    }
    const weight = criterion.weight ?? 1;
    totalScore += response.score * weight;
    maxScore += criterion.maxScore * weight;
  }

  const { percentage, mention, scoreOn4 } = convertRawScore(
    conversionTable,
    totalScore,
    maxScore,
    section.criteria.length,
  );

  return { totalScore, maxScore, percentage, mention, scoreOn4 };
}
